// ngx-marquee/esm2020/lib/observables/from-intersection-observer.mjs
import { Observable, Subject } from "rxjs";
import { debounceTime, filter } from "rxjs/operators";
export var IntersectionStatus;
(function (IntersectionStatus) {
    IntersectionStatus["Visible"] = "Visible";
    IntersectionStatus["Pending"] = "Pending";
    IntersectionStatus["NotVisible"] = "NotVisible";
})(IntersectionStatus || (IntersectionStatus = {}));
export const fromIntersectionObserver = (element, config, debounce = 0) => new Observable(subscriber => {
    const subject$ = new Subject();
    const intersectionObserver = new IntersectionObserver((entries, observer) => {
        entries.forEach(entry => {
            subject$.next({ entry, observer });
        });
    }, config);
    subject$.subscribe(() => {
        subscriber.next(IntersectionStatus.Pending);
    });
    subject$.pipe(debounceTime(debounce), filter(Boolean))
        .subscribe(async ({ entry, observer }) => {
        const isEntryVisible = await isVisible(entry.target);
        if (isEntryVisible) {
            subscriber.next(IntersectionStatus.Visible);
        }
        else {
            subscriber.next(IntersectionStatus.NotVisible);
        }
    });
    intersectionObserver.observe(element);
    return {
        unsubscribe() {
            intersectionObserver.disconnect();
            subject$.unsubscribe();
        }
    };
});
async function isVisible(element) {
    return new Promise(resolve => {
        const observer = new IntersectionObserver(([entry]) => {
            resolve(entry.isIntersecting);
            observer.disconnect();
        });
        observer.observe(element);
    });
}
