// ngx-marquee/esm2020/lib/ngx-marquee.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NgxMarqueeComponent } from './ngx-marquee.component';
import { IntersectionObserverDirective } from './directives/intersection-observer.directive';
import * as i0 from "@angular/core";
export class NgxMarqueeModule {
}
NgxMarqueeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.0", ngImport: i0, type: NgxMarqueeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgxMarqueeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.0", ngImport: i0, type: NgxMarqueeModule, declarations: [NgxMarqueeComponent,
        IntersectionObserverDirective], imports: [CommonModule], exports: [NgxMarqueeComponent] });
NgxMarqueeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.0", ngImport: i0, type: NgxMarqueeModule, imports: [[
            CommonModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.0", ngImport: i0, type: NgxMarqueeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        NgxMarqueeComponent,
                        IntersectionObserverDirective
                    ],
                    imports: [
                        CommonModule
                    ],
                    exports: [
                        NgxMarqueeComponent
                    ]
                }]
        }] });
